File: typescript/examples/quant-tier/stream-bids-snapshot/stream_bids_snapshot.ts
import { fileURLToPath } from "node:url";
import { parseArgs } from "node:util";

import type { ApiResponse, BuyOrderItem } from "cs2cap";
import { BidsApi, BuyOrderItemFromJSON } from "cs2cap";

import { formatApiError } from "../../_shared/api.js";
import {
  buildConfiguration,
  loadApiKey,
  requireRealApiKey,
} from "../../_shared/auth.js";
import { parseIntegerOption } from "../../_shared/cli.js";
import { formatMoney, renderTable } from "../../_shared/render.js";
import { decodeLines } from "../../_shared/streaming.js";

const SCRIPT_PATH = fileURLToPath(import.meta.url);

function readOptions(): { maxLines: number; top: number; provider?: string } {
  const { values } = parseArgs({
    options: {
      "max-lines": { type: "string" },
      top: { type: "string" },
      provider: { type: "string" },
    },
  });

  return {
    maxLines: parseIntegerOption(values["max-lines"], 5000, "--max-lines"),
    top: parseIntegerOption(values.top, 15, "--top"),
    provider: values.provider,
  };
}

async function main(): Promise<number> {
  const args = readOptions();

  let bearerToken: string;
  try {
    bearerToken = loadApiKey("CS2C_API_KEY", SCRIPT_PATH);
    requireRealApiKey(bearerToken);
  } catch (error) {
    console.error(error instanceof Error ? error.message : String(error));
    return 1;
  }

  const configuration = buildConfiguration(bearerToken);
  const bidsApi = new BidsApi(configuration);
  const collected: BuyOrderItem[] = [];
  let processedLines = 0;
  let response: ApiResponse<string> | undefined;

  try {
    response = await bidsApi.streamBidsSnapshotV1BidsPostRaw();
    for await (const line of decodeLines(response.raw.body)) {
      const item = BuyOrderItemFromJSON(JSON.parse(line));
      processedLines += 1;

      if (args.provider && item.provider !== args.provider) {
        if (args.maxLines > 0 && processedLines >= args.maxLines) {
          break;
        }
        continue;
      }

      collected.push(item);
      if (args.maxLines > 0 && processedLines >= args.maxLines) {
        break;
      }
    }
  } catch (error) {
    console.error(await formatApiError(error));
    return 1;
  } finally {
    if (response?.raw.body) {
      await response.raw.body.cancel().catch(() => undefined);
    }
  }

  if (collected.length === 0) {
    console.log("No rows were collected from the processed bids snapshot window.");
    return 0;
  }

  collected.sort((left, right) => right.highestBid - left.highestBid);
  const topRows = collected.slice(0, args.top);

  console.log("Snapshot summary:");
  console.log(`- processed_lines: ${processedLines}`);
  console.log(`- collected_rows: ${collected.length}`);
  console.log(`- unique_providers: ${new Set(collected.map((item) => item.provider)).size}`);
  if (args.provider) {
    console.log(`- provider_filter: ${args.provider}`);
  }
  console.log();

  console.log(
    renderTable(
      ["provider", "item_id", "market_hash_name", "highest_bid", "num_bids"],
      topRows.map((item) => [
        item.provider,
        item.itemId,
        item.marketHashName,
        formatMoney(item.highestBid),
        item.numBids,
      ]),
    ),
  );
  return 0;
}

void main().then((code) => {
  process.exitCode = code;
});

File: python/examples/quant-tier/stream-bids-snapshot/stream_bids_snapshot.py
import argparse
import json
import sys
from pathlib import Path

import cs2cap
from cs2cap.rest import ApiException

EXAMPLES_ROOT = Path(__file__).resolve().parents[2]
if str(EXAMPLES_ROOT) not in sys.path:
    sys.path.insert(0, str(EXAMPLES_ROOT))

from _shared.auth import build_configuration, load_api_key, require_real_api_key  # noqa: E402
from _shared.render import format_money, render_table  # noqa: E402
from _shared.streaming import decode_lines  # noqa: E402


def build_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser(
        description="Stream the /v1/bids NDJSON snapshot stream.",
    )
    parser.add_argument(
        "--max-lines",
        type=int,
        default=5000,
        help="Maximum NDJSON rows to process. Use 0 to consume the full stream.",
    )
    parser.add_argument(
        "--top",
        type=int,
        default=15,
        help="Number of strongest bid rows to print after processing.",
    )
    parser.add_argument(
        "--provider",
        default=None,
        help="Optional local provider filter applied after decoding each row.",
    )
    return parser


def main() -> int:
    args = build_parser().parse_args()

    try:
        bearer_token = load_api_key("CS2C_API_KEY", start=Path(__file__))
        require_real_api_key(bearer_token)
    except RuntimeError as exc:
        print(str(exc), file=sys.stderr)
        return 1

    configuration = build_configuration(bearer_token)
    collected: list[cs2cap.BuyOrderItem] = []
    processed_lines = 0
    response = None

    try:
        with cs2cap.ApiClient(configuration) as client:
            bids_api = cs2cap.BidsApi(client)
            response = bids_api.stream_bids_snapshot_v1_bids_post_without_preload_content()

            for line in decode_lines(response):
                payload = json.loads(line)
                item = cs2cap.BuyOrderItem.from_dict(payload)
                processed_lines += 1

                if args.provider and item.provider != args.provider:
                    if args.max_lines and processed_lines >= args.max_lines:
                        break
                    continue

                collected.append(item)
                if args.max_lines and processed_lines >= args.max_lines:
                    break

    except ApiException as exc:
        print(f"API request failed: {exc}", file=sys.stderr)
        return 1
    finally:
        if response is not None:
            response.release_conn()
            response.close()

    if not collected:
        print("No rows were collected from the processed bids snapshot window.")
        return 0

    collected.sort(key=lambda item: item.highest_bid, reverse=True)
    top_rows = collected[: args.top]

    print("Snapshot summary:")
    print(f"- processed_lines: {processed_lines}")
    print(f"- collected_rows: {len(collected)}")
    print(f"- unique_providers: {len({item.provider for item in collected})}")
    if args.provider:
        print(f"- provider_filter: {args.provider}")
    print()

    rows = [
        [
            item.provider,
            str(item.item_id),
            item.market_hash_name,
            format_money(item.highest_bid),
            str(item.num_bids),
        ]
        for item in top_rows
    ]
    print(
        render_table(["provider", "item_id", "market_hash_name", "highest_bid", "num_bids"], rows)
    )  # noqa: E501
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
